helix-core/src/main/java/org/apache/helix/controller/stages/MessageGenerationPhase.java

```diff
@@ -220,7 +220,8 @@ private void generateMessage(final Resource resource, final BaseControllerDataPr
                 generateCancellationMessageForPendingMessage(desiredState, currentState, nextState, pendingMessage,
                     manager, resource, partition, sessionIdMap, instanceName, stateModelDef,
                     cancellationMessage, isCancellationEnabled);
-          } else {
+          }
+          if (pendingMessage == null || currentState.equalsIgnoreCase(pendingMessage.getToState())) {
```

Comment:
When "pendingMessage != null" but "currentState.equalsIgnoreCase(pendingMessage.getToState())", the cancellation message will be replaced with the new ST message right? I think it works. But would it be better to send both of the cancellation message and the new one?

If we do this, will the participant discards the cancellation message, or proceed with both of them? It would be safer from Helix perspective since the pending message will be removed for sure with cancellation. But it could be complicated from the participant perspective since I think we might need some sync control on the participant side.

Reply:
Actually if looking at the generateCancellationMessageForPendingMessage logic, there is

`else if (currentState.equalsIgnoreCase(pendingState)) {
        LogUtil.logInfo(logger, _eventId,
            "Message hasn't been removed for " + instanceName + " to transit " + resource
                .getResourceName() + "." + partition.getPartitionName() + " to "
                + pendingState + ", desiredState: " + desiredState + ", isRelay: " + pendingMessage.isRelayMessage());
      }`

So actually when currentState equals pendingState, there is no cancellation message generated.

Reply:
Then it means the decision of canceling or not is purely in the generateCancellationMessageForPendingMessage(). And we don't have the option to send both of the messages for now.

In this case, it would be cleaner to do the following, I guess.
```
if (pendingMessage != null) {
            message =
                generateCancellationMessageForPendingMessage(....);
}
if (message == null) {
            // Create new state transition message
            message = createStateTransitionMessage(...);
}
```

Reply:
I was thinking if there is a way to avoid the same condition being checked inside generateCancellationMessageForPendingMessage and here. But it seems that there is no easy way.

However, to avoid the possible unexpected overwrite (for instance, if we change the private method logic later), I think we can change as following,

Option one,
`if (message == null && (pendingMessage == null || currentState.equalsIgnoreCase(pendingMessage.getToState()))) {`
Which means if there is no cancellation message and the the pending message has been done then we generate new ST.

Option two,
Remove generateCancellationMessageForPendingMessage and move all if-else conditions back to the caller methods. Because you do need to handle those conditions with different logic. That seems to be a clearer logic.

Reply:
GenerateCancellationMessageForPendingMessage is used in different places. Doesn't look good to move that function back. We can guard it with "message == null" for now.

Reply:
I think the other place will never hit the other 2 conditions. But not 100% sure.
Anyway, both options work.